compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -285,12 +291,23 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     };
                     check_equal(self, location, f_ty);
                 }
-                ty::Generator(_, substs, _) => {
-                    let substs = substs.as_generator();
-                    let Some(f_ty) = substs.upvar_tys().nth(f.as_usize()) else {
+                ty::Generator(def_id, _, _) => {
+                    let var = parent_ty.variant_index.unwrap_or(VariantIdx::from_u32(0));
+                    let Some(layout) = self.tcx.generator_layout(*def_id) else {
+                        self.fail(location, format!("No generator layout for {:?}", parent_ty));
+                        return;
+                    };
```

Comment:
The query cycle is because I messed up here, instead of going through a query, it should be getting the generator layout directly through the MIR body. I'll fix it soon.

Reply:
How do you plan to do that? The generator `resume` call could be inlined